### PR TITLE
Extending KEVM configuration with `<wordStackSize>`

### PR DESCRIPTION
**Pull request by PetarMax**
This PR introduces a `<wordStackSize>` cell into the EVM configuration, which holds the size of the word stack. This should remove the need to continuously compute the size of the word stack, which is both recursive and cannot be done using the LLVM backend if any of the elements of the word stack are symbolic.

What needs to be done carefully in this change is to preserve the invariant that having
```
<wordStack> WS </wordStack>
<wordStackSize> WSSize </wordStackSize>
```
means that `WSSize ==Int #sizeWordStack(WS)`.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md [hunk 1]
```diff
@@ -356,11 +359,11 @@ The `#next [_]` operator initiates execution by:
...
 
 ```k
-    syntax Bool ::= #stackUnderflow ( WordStack , OpCode ) [symbol(#stackUnderflow), macro]
-                  | #stackOverflow  ( WordStack , OpCode ) [symbol(#stackOverflow), macro]
+    syntax Bool ::= #stackUnderflow ( Int , OpCode ) [symbol(#stackUnderflow), macro]
+                  | #stackOverflow  ( Int , OpCode ) [symbol(#stackOverflow), macro]
  // ---------------------------------------------------------------------------------------
-    rule #stackUnderflow(WS, OP:OpCode) => #sizeWordStack(WS) <Int #stackNeeded(OP)
-    rule #stackOverflow (WS, OP) => #sizeWordStack(WS) +Int #stackDelta(OP) >Int 1024
+    rule #stackUnderflow(WSSize, OP:OpCode) => WSSize <Int #stackNeeded(OP)
```
Comment (anvacaru) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md:365]: Could you change this to all caps for consistency?
```suggestion
    rule #stackUnderflow(WSSIZE, OP:OpCode) => WSSize <Int #stackNeeded(OP)
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 2]
```diff
@@ -81,7 +88,8 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( WS ) <=Int 1023 )
+     andBool ( 0 <=Int WSSize )
+     andBool ( WSSize <=Int 1023 )
```
Comment (anvacaru) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:92]: Do we need a `#rangeWSSize`? I see these constraints are only used in `optimizations.md`.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 3]
```diff
@@ -335,7 +370,8 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( WS ) <=Int 1023 )
+     andBool ( 2 <=Int WSSize )
+     andBool ( WSSize <=Int 1023 )
```
Comment (anvacaru) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:374]: Oh, and the lower bound is different depending on the opcode.

**Review by anvacaru — APPROVED** [review]
Looks good to me. I left some questions. Do you also have a performance profiling? I'm curious if this adds any performance improvements.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md (hunk 1 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md]: Yes, good idea, I have done this.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md (hunk 2 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md]: No, we can't, it's a bit dynamic.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md (hunk 3 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md]: For example.

**Comment by PetarMax** [comment]
On one of our engagement test suites, the performance is as follows (with parallel factor 6):

|     | wssize | master | speedup |
|----:|-------:|-------:|--------:|
|  T1 |    170 |    175 | -2.86%  |
|  T2 |    133 |    139 | -4.32%  |
|  T3 |    104 |    113 | -7.96%  |
|  T4 |    316 |    342 | -7.60%  |
|  T5 |   1573 |   1699 | -7.42%  |
|  T6 |    272 |    339 | -19.76% |
|  T7 |    165 |    200 | -17.50% |
|  T8 |    104 |    111 | -6.31%  |
|  T9 |   4744 |   4901 | -3.20%  |
| T10 |    118 |    128 | -7.81%  |
| T11 |    927 |    998 | -7.11%  |

so we can see that there is a speed-up throughout, with varying extent.

**Comment by PetarMax** [comment]
On another engagement-related test suite, the times are as follows:

|     | wssize | master | speedup |
|----:|-------:|-------:|--------:|
|  T1 |     42 |     42 | 0.00%   |
|  T2 |     34 |     34 | 0.00%   |
|  T3 |     39 |     45 | -13.33% |
|  T4 |     41 |     47 | -12.77% |
|  T5 |    262 |    365 | -28.22% |
|  T6 |    477 |    581 | -17.90% |
|  T7 |    131 |    180 | -27.22% |

with a higher speed-up on average.

I think that this depends on just how much stack manipulation is happening.

**Pull request closed without merging**